### ISSUE TITLE
failed to terminate or stop with PNS executor

### ISSUE THREAD

**Issue by sigfridh** [argoproj/argo#4059]
## Summary 

When i start a workflow with a PNS executor im unable to stop or teminate it, for this i run a simple sleep task.
argo version is 2.10.2 (also try with 2.9) and my container runtime is containerd

I think its due to the fact that the wait container is terminating:

``
time="2020-09-17T13:23:10.705Z" level=info msg="Starting Workflow Executor" version=v2.10.2
time="2020-09-17T13:23:10.709Z" level=info msg="Creating PNS executor (namespace: argo, pod: lovely-dog, pid: 6, hasOutputs: false)"
time="2020-09-17T13:23:10.709Z" level=info msg="Executor (version: v2.10.2, build_date: 2020-09-14T17:38:28Z) initialized (pod: argo/lovely-dog) with template:\n{\"name\":\"dogprint\",\"arguments\":{},\"inputs\":{},\"outputs\":{},\"metadata\":{},\"container\":{\"name\":\"main\",\"image\":\"us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1\",\"command\":[\"sh\",\"-c\"],\"args\":[\"sleep 200\"],\"resources\":{}},\"serviceAccountName\":\"argo\"}"
time="2020-09-17T13:23:10.709Z" level=info msg="Waiting on main container"
time="2020-09-17T13:23:10.866Z" level=info msg="main container started with container ID: cad3533a770168a7dbdbb8387e98863ff8cf294b11891934b92be44843ea305b"
time="2020-09-17T13:23:10.866Z" level=info msg="Starting annotations monitor"
time="2020-09-17T13:23:10.868Z" level=info msg="Starting deadline monitor"
time="2020-09-17T13:23:10.869Z" level=info msg="containerID cri-containerd-cad3533a770168a7dbdbb8387e98863ff8cf294b11891934b92be44843ea305b mapped to pid 17"
time="2020-09-17T13:23:10.869Z" level=warning msg="Ignoring wait failure: Failed to determine pid for containerID cad3533a770168a7dbdbb8387e98863ff8cf294b11891934b92be44843ea305b: container may have exited too quickly. Process assumed to have completed"
time="2020-09-17T13:23:10.869Z" level=info msg="Main container completed"
time="2020-09-17T13:23:10.869Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2020-09-17T13:23:10.869Z" level=info msg="Capturing script exit code"
time="2020-09-17T13:23:10.869Z" level=info msg="Getting exit code of cad3533a770168a7dbdbb8387e98863ff8cf294b11891934b92be44843ea305b"
time="2020-09-17T13:23:10.869Z" level=info msg="Annotations monitor stopped"
time="2020-09-17T13:23:10.872Z" level=info msg="No output parameters"
time="2020-09-17T13:23:10.872Z" level=info msg="No output artifacts"
time="2020-09-17T13:23:10.872Z" level=info msg="Killing sidecars"
time="2020-09-17T13:23:10.876Z" level=info msg="Alloc=5931 TotalAlloc=14287 Sys=70080 NumGC=4 Goroutines=9"``

this log:

"Ignoring wait failure: Failed to determine pid for containerID cad3533a770168a7dbdbb8387e98863ff8cf294b11891934b92be44843ea305b: container may have exited too quickly. Process assumed to have completed"

Any ideas?

Also i check with kubelet executor, in this case the wait container is ok, the kill command is forwarded, but pod is not killed (kill timeout

Thx
sigfrid

**Comment by sigfridh** [comment]
Maybe th 'cri-containerd-' prefix? like in the crio proble: 
https://github.com/lippertmarkus/argo/commit/4a4f8300f6d9dc463f73dafc0c62b5b29e320be3

**Comment by alexec** [comment]
@sarabala1979 some kind of sleep work-around?

**Comment by sarabala1979** [comment]
looks like it is know issue in PNS executor. If main container finished quickly, the wait container is not able to secure the FS.
Can you try using the `emptydir` as a workaround for above issue?
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: base-image-out-art-
spec:
  entrypoint: base-image-out-art
  volumes:
  - name: workdir
    emptyDir: {}
  templates:
  - name: base-image-out-art
    container:
      image: alpine:latest
      command: [echo, hello]
      volumeMounts:
      - name: workdir
        mountPath: /mnt/vol
```
Can you provide what error message you are getting on `stop` and `terminate`?

**Comment by gmfmi** [comment]
Hi everyone, I got the exact same issue.

@sarabala1979, I tried with your empty-dir example. I tweaked it a little bit to be able to send stop and terminate events (switched the alpine command from `echo` to `sleep 30`). Here are the outputs:

```sh
# From the workflow-controller
time="2020-09-21T07:49:19Z" level=info msg="Processing workflow" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:19Z" level=info msg="Updated phase  -> Running" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:19Z" level=info msg="Pod node base-image-out-art-ng5hm initialized Pending" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:19Z" level=info msg="Created pod: base-image-out-art-ng5hm (base-image-out-art-ng5hm)" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:19Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=19943 workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:19Z" level=info msg="Processing workflow" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:19Z" level=info msg="Updating node base-image-out-art-ng5hm message: ContainerCreating"
time="2020-09-21T07:49:19Z" level=info msg="Skipped pod base-image-out-art-ng5hm (base-image-out-art-ng5hm) creation: already exists" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:19Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=19946 workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:19Z" level=info msg="Processing workflow" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:19Z" level=info msg="Skipped pod base-image-out-art-ng5hm (base-image-out-art-ng5hm) creation: already exists" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:20Z" level=info msg="insignificant pod change" key=argo/base-image-out-art-ng5hm
time="2020-09-21T07:49:20Z" level=info msg="Processing workflow" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:20Z" level=info msg="Skipped pod base-image-out-art-ng5hm (base-image-out-art-ng5hm) creation: already exists" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:24Z" level=info msg="Processing workflow" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:24Z" level=info msg="Updating node base-image-out-art-ng5hm status Pending -> Running"
time="2020-09-21T07:49:24Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=19969 workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:24Z" level=info msg="Processing workflow" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:25Z" level=info msg="Processing workflow" namespace=argo workflow=base-image-out-art-ng5hm

[### STOP EVENT ###]
time="2020-09-21T07:49:35Z" level=info msg="Processing workflow" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:35Z" level=info msg="Applying shutdown deadline for pod base-image-out-art-ng5hm" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:35Z" level=info msg="Updating execution control of base-image-out-art-ng5hm: {\"deadline\":\"0001-01-01T00:00:00Z\"}" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:35Z" level=info msg="Signalling base-image-out-art-ng5hm of updates" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:35Z" level=info msg="https://10.96.0.1:443/api/v1/namespaces/argo/pods/base-image-out-art-ng5hm/exec?command=sh&command=-c&command=kill+-s+USR2+%24%28pidof+argoexec%29&container=wait&stderr=true&stdout=true&tty=false"
time="2020-09-21T07:49:35Z" level=info msg="insignificant pod change" key=argo/base-image-out-art-ng5hm
time="2020-09-21T07:49:35Z" level=warning msg="Signal command failed: unable to upgrade connection: container not found (\"wait\")" namespace=argo workflow=base-image-out-art-ng5hm

[### TERMINATE EVENT ###]
time="2020-09-21T07:49:41Z" level=info msg="Processing workflow" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:41Z" level=info msg="Applying shutdown deadline for pod base-image-out-art-ng5hm" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:41Z" level=info msg="Updating execution control of base-image-out-art-ng5hm: {\"deadline\":\"0001-01-01T00:00:00Z\"}" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:41Z" level=info msg="Signalling base-image-out-art-ng5hm of updates" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:41Z" level=info msg="https://10.96.0.1:443/api/v1/namespaces/argo/pods/base-image-out-art-ng5hm/exec?command=sh&command=-c&command=kill+-s+USR2+%24%28pidof+argoexec%29&container=wait&stderr=true&stdout=true&tty=false"
time="2020-09-21T07:49:41Z" level=warning msg="Signal command failed: unable to upgrade connection: container not found (\"wait\")" namespace=argo workflow=base-image-out-art-ng5hm

time="2020-09-21T07:49:54Z" level=info msg="Processing workflow" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:54Z" level=info msg="Updating node base-image-out-art-ng5hm status Running -> Succeeded"
time="2020-09-21T07:49:54Z" level=info msg="Updated phase Running -> Succeeded" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:54Z" level=info msg="Marking workflow completed" namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:54Z" level=info msg="Checking daemoned children of " namespace=argo workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:54Z" level=info msg="Workflow update successful" namespace=argo phase=Succeeded resourceVersion=20046 workflow=base-image-out-art-ng5hm
time="2020-09-21T07:49:54Z" level=info msg="Labeled pod argo/base-image-out-art-ng5hm completed"
```

```sh
# from "base-image-out-art-ng5hm" (wait container)
time="2020-09-21T07:49:21.384Z" level=info msg="Starting Workflow Executor" version=v2.9.3
time="2020-09-21T07:49:21.391Z" level=info msg="Creating PNS executor (namespace: argo, pod: base-image-out-art-ng5hm, pid: 6, hasOutputs: false)"
time="2020-09-21T07:49:21.391Z" level=info msg="Executor (version: v2.9.3, build_date: 2020-07-18T19:11:19Z) initialized (pod: argo/base-image-out-art-ng5hm) with template:\n{\"name\":\"base-image-out-art\",\"arguments\":{},\"inputs\":{},\"outputs\":{},\"metadata\":{},\"container\":{\"name\":\"\",\"image\":\"alpine:latest\",\"command\":[\"sleep\",\"30\"],\"resources\":{},\"volumeMounts\":[{\"name\":\"workdir\",\"mountPath\":\"/mnt/vol\"}]}}"
time="2020-09-21T07:49:21.391Z" level=info msg="Waiting on main container"
time="2020-09-21T07:49:23.367Z" level=info msg="main container started with container ID: 14480c4b36d996f2204e772b5503c822bc12dfe9c7c736b8098f2df1339cc4e3"
time="2020-09-21T07:49:23.369Z" level=info msg="Starting annotations monitor"
time="2020-09-21T07:49:23.411Z" level=info msg="containerID cri-containerd-14480c4b36d996f2204e772b5503c822bc12dfe9c7c736b8098f2df1339cc4e3 mapped to pid 17"
time="2020-09-21T07:49:23.411Z" level=warning msg="Ignoring wait failure: Failed to determine pid for containerID 14480c4b36d996f2204e772b5503c822bc12dfe9c7c736b8098f2df1339cc4e3: container may have exited too quickly. Process assumed to have completed"
time="2020-09-21T07:49:23.412Z" level=info msg="Main container completed"
time="2020-09-21T07:49:23.412Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2020-09-21T07:49:23.412Z" level=info msg="Capturing script exit code"
time="2020-09-21T07:49:23.412Z" level=info msg="Getting exit code of 14480c4b36d996f2204e772b5503c822bc12dfe9c7c736b8098f2df1339cc4e3"
time="2020-09-21T07:49:23.412Z" level=info msg="Annotations monitor stopped"
time="2020-09-21T07:49:23.411Z" level=info msg="Starting deadline monitor"
time="2020-09-21T07:49:23.414Z" level=info msg="Deadline monitor stopped"
time="2020-09-21T07:49:23.418Z" level=info msg="No output parameters"
time="2020-09-21T07:49:23.418Z" level=info msg="No output artifacts"
time="2020-09-21T07:49:23.418Z" level=info msg="Killing sidecars"
time="2020-09-21T07:49:23.423Z" level=info msg="Alloc=6628 TotalAlloc=15340 Sys=70080 NumGC=4 Goroutines=8"
```

Hope it will be useful, I can get more outputs if needed.

**Comment by alexec** [comment]
@sarabala1979 can you please link the know issue you mentioned?

**Comment by gmfmi** [comment]
@alexec I did not dig into the source code but I do not think this issue is related to the emptyDir one. The container launched runs for more than 30 seconds so it is probably not a matter of starting/terminating to quickly.

**Comment by alexec** [comment]
@gmfmi you logs seem to show the container exiting after 2s.

```
time="2020-09-21T07:49:23.411Z" level=info msg="containerID cri-containerd-14480c4b36d996f2204e772b5503c822bc12dfe9c7c736b8098f2df1339cc4e3 mapped to pid 17"
time="2020-09-21T07:49:23.411Z" level=warning msg="Ignoring wait failure: Failed to determine pid for containerID 14480c4b36d996f2204e772b5503c822bc12dfe9c7c736b8098f2df1339cc4e3: container may have exited too quickly. Process assumed to have completed"
```

Can you please double check the logs of the main container?

**Comment by gmfmi** [comment]
@alexec I did not kept the logs of the "main" container but it was running for 30 secs (running status on Kubectl CLI and Argo UI) . If you look at the workflow controller logs, the pod starts at `2020-09-21T07:49:19Z` and ends at `2020-09-21T07:49:54Z`, so the hole process took 35 secs in total.

To me, it's much more an issue about the "wait" container that fails to watch/wait properly the "main" container than a misconfiguration of the "main". I think that's why @sigfridh is talking about a non-matching prefix issue.

**Comment by alexec** [comment]
@gmfmi is the past, users that has had this problem has had it because the main container exited too quickly. I think it therefore unlikely (but not impossible) that it is an issue with the wait container. 

The section of logs you're provided for the wait container is <30s.

While the pod may run for 30s, it is possible for containers within the pod to run for less time, and this can be confirmed by either timestamps on the main log, or from the pod status `/status/contaninerStatuses`.

I want to be 100% sure that the main container ran for >=30s. Could I please ask you to provide either the main container logs or the full pod status.

Thank you.

**Comment by gmfmi** [comment]
@alexec, thank you for the context.

About the main container logs, it is empty because I only made a `sleep 30`. If it can be helpful, I can run a new test with (for example) something that prints out the date every seconds.

Now, by analysing the pod's description, we can see that the "main" container runs exactly 30s. I copied/pasted the entire output, it may helps you find something interesting.

```sh
$ kubectl -n argo describe pod base-image-out-art-ng5hm

Name:         base-image-out-art-ng5hm
Namespace:    argo
Priority:     0
Node:         node1/192.168.56.11
Start Time:   Mon, 21 Sep 2020 07:49:19 +0000
Labels:       workflows.argoproj.io/completed=true
              workflows.argoproj.io/workflow=base-image-out-art-ng5hm
Annotations:  cni.projectcalico.org/podIP: 10.244.166.149/32
              cni.projectcalico.org/podIPs: 10.244.166.149/32
              workflows.argoproj.io/execution: {"deadline":"0001-01-01T00:00:00Z"}
              workflows.argoproj.io/node-name: base-image-out-art-ng5hm
              workflows.argoproj.io/template:
                {"name":"base-image-out-art","arguments":{},"inputs":{},"outputs":{},"metadata":{},"container":{"name":"","image":"alpine:latest","command...
Status:       Succeeded
IP:           10.244.166.149
IPs:
  IP:           10.244.166.149
Controlled By:  Workflow/base-image-out-art-ng5hm
Containers:
  wait:
    Container ID:  containerd://bb9b3925deacef264c760d1320a88263aa3ca6a1849f8dec13ce078e69e4723d
    Image:         kast-registry:30005/argoproj/argoexec:v2.9.3
    Image ID:      sha256:929b13037cf471c15db48778af2436caf0be4040f649a261fe824e8524f53eab
    Port:          <none>
    Host Port:     <none>
    Command:
      argoexec
      wait
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Mon, 21 Sep 2020 07:49:21 +0000
      Finished:     Mon, 21 Sep 2020 07:49:23 +0000
    Ready:          False
    Restart Count:  0
    Environment:
      ARGO_POD_NAME:                    base-image-out-art-ng5hm (v1:metadata.name)
      ARGO_CONTAINER_RUNTIME_EXECUTOR:  pns
    Mounts:
      /argo/podmetadata from podmetadata (rw)
      /mainctrfs/mnt/vol from workdir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-8kl67 (ro)
  main:
    Container ID:  containerd://14480c4b36d996f2204e772b5503c822bc12dfe9c7c736b8098f2df1339cc4e3
    Image:         alpine:latest
    Image ID:      docker.io/library/alpine@sha256:185518070891758909c9f839cf4ca393ee977ac378609f700f60a771a2dfe321
    Port:          <none>
    Host Port:     <none>
    Command:
      sleep
      30
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Mon, 21 Sep 2020 07:49:22 +0000
      Finished:     Mon, 21 Sep 2020 07:49:52 +0000
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /mnt/vol from workdir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-8kl67 (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  podmetadata:
    Type:  DownwardAPI (a volume populated by information about the pod)
    Items:
      metadata.annotations -> annotations
  workdir:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  default-token-8kl67:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-8kl67
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason     Age        From               Message
  ----    ------     ----       ----               -------
  Normal  Scheduled  <unknown>  default-scheduler  Successfully assigned argo/base-image-out-art-ng5hm to node1
  Normal  Pulled     2d9h       kubelet, node1     Container image "kast-registry:30005/argoproj/argoexec:v2.9.3" already present on machine
  Normal  Created    2d9h       kubelet, node1     Created container wait
  Normal  Started    2d9h       kubelet, node1     Started container wait
  Normal  Pulling    2d9h       kubelet, node1     Pulling image "alpine:latest"
  Normal  Pulled     2d9h       kubelet, node1     Successfully pulled image "alpine:latest"
  Normal  Created    2d9h       kubelet, node1     Created container main
  Normal  Started    2d9h       kubelet, node1     Started container main
```

**Comment by alexec** [comment]
I am now convinced. Can you please find me on Slack to discus maybe Zoom. I'm in PST.

**Comment by gmfmi** [comment]
@alexec, sorry for the delay I am mainly doing that on my free time. I send you a DM on Slack. UTC+2 for me...

**Comment by alexec** [comment]
See #4289

**Comment by cy-zheng** [comment]
Root cause could be #4302.

**Comment by gmfmi** [comment]
Hi @cy-zheng, I would love to get these records but I won't get any access to my dev environment (I left the company). To me, it is exactly the same issue and I think that's also what @sigfridh was pointing out.

I think your solution will also close this issue, thanks a lot!

**Comment by alexec** [comment]
I think this may be fixed by #4253. Can you please try `argoproj/argoexec:latest`?

**Comment by gmfmi** [comment]
> I think this may be fixed by #4253. Can you please try `argoproj/argoexec:latest`?

I will give it a try soon, expect an answer in the next 24 hours!

**Comment by gmfmi** [comment]
Just tested the latest release: everything works as expected. Thank you @cy-zheng ! @alexec, feel free to close the issue.
